### PR TITLE
feat(heuristics): add SimilarProjectAnalyzer to detect structural similarity across packages from same maintainer

### PR DESCRIPTION
**Pull request by AmineRaouane**
## Summary

This PR adds a new heuristic analyzer called `SimilarProjectAnalyzer`. It checks whether a PyPI package has a similar file/folder structure to other packages maintained by the same user. This helps in identifying potentially malicious packages that replicate existing structures.

## Description of changes

- Created a new analyzer: `SimilarProjectAnalyzer`.
- The analyzer fetches the list of maintainers of the target package and retrieves other packages published by those maintainers.
- For each package, it computes a normalized structure hash from its sdist tarball and compares it to the structure hash of the target package.
- If any match is found, the heuristic fails, flagging potential structural duplication.
- Added this analyzer to the `heuristics.py` registry.
- Modified `detect_malicious_metadata_check.py` to include and utilize the new heuristic.
- Added test cases to validate the functionality and edge cases of the analyzer.

## Related issues

None

- [x] I have reviewed the [contribution guide](../CONTRIBUTING.md).
- [x] My PR title and commits follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) convention.
- [x] My commits include the "Signed-off-by" line.
- [x] I have signed my commits following the instructions provided by [GitHub](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits). Note that we run [GitHub's commit verification](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) tool to check the commit signatures. A green `verified` label should appear next to **all** of your commits on GitHub.
- [x] I have updated the relevant documentation, if applicable.
- [x] I have tested my changes and verified they work as expected.

### CODE REVIEW

File: src/macaron/malware_analyzer/pypi_heuristics/metadata/similar_projects.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+            if package_hash is None:
+                logger.info("Package does not have a sdist.")
+                continue
+            if package_hash == target_hash:
+                return HeuristicResult.FAIL, {
+                    "similar_package": package,
+                }
+        return HeuristicResult.PASS, {}
+
+    def get_maintainers(self, package_name: str) -> list[str]:
```
Comment (art1f1c3R) [inline src/macaron/malware_analyzer/pypi_heuristics/metadata/similar_projects.py:75]: I believe this functionality is offered by `PyPIRegistry.get_maintainers_of_package`.

File: src/macaron/malware_analyzer/pypi_heuristics/metadata/similar_projects.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+        maintainers = self.get_maintainers(package_name)
+        for user in maintainers:
+            user_packages = self.get_packages_by_user(user)
+            similar_projects.extend(user_packages)
+        # Remove the target package from the list of similar projects.
+        similar_projects_set = set(similar_projects)
+        similar_projects_set.discard(package_name)
+        return list(similar_projects_set)
+
+    def fetch_sdist_url(self, package_name: str, version: str | None = None) -> str:
```
Comment (art1f1c3R) [inline src/macaron/malware_analyzer/pypi_heuristics/metadata/similar_projects.py:139]: I wonder if it would be possible to create a `PyPIPackageJsonAsset` object so you may use its function `get_sourcecode_url` for this?

File: src/macaron/malware_analyzer/pypi_heuristics/metadata/similar_projects.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        Raises
+        ------
+            ValueError: If the sdist URL cannot be fetched or the package structure cannot be hashed.
+        """
+        sdist_url = self.fetch_sdist_url(package_name)
+        if not sdist_url:
+            return None
+
+        try:
+            response = requests.get(sdist_url, stream=True, timeout=10)
```
Comment (art1f1c3R) [inline src/macaron/malware_analyzer/pypi_heuristics/metadata/similar_projects.py:205]: Similar here with using `util.py`

File: src/macaron/slsa_analyzer/checks/detect_malicious_metadata_check.py [hunk 4]
```diff
@@ -381,6 +383,10 @@ def run_check(self, ctx: AnalyzeContext) -> CheckResultData:
         failed({Heuristics.CLOSER_RELEASE_JOIN_DATE.value}),
         forceSetup.
 
+    % Package released that is similar to other packages maintained by the same maintainer.
```
Comment (art1f1c3R) [inline src/macaron/slsa_analyzer/checks/detect_malicious_metadata_check.py:386]: Could you walk me through the rationale of why we should combine `SIMILAR_PROJECTS` failing with the `forceSetup` and `quickUndetailed` rules and why these rules together are a malicious indicator with `HIGH` confidence?

File: src/macaron/malware_analyzer/pypi_heuristics/metadata/similar_projects.py (hunk 1 above)
Comment (AmineRaouane) [reply src/macaron/malware_analyzer/pypi_heuristics/metadata/similar_projects.py:75]: Correct. That is what I used first, but @behnazh-w  told me that get_maintainers_of_package is not working anymore because PYPI blocks it so I rewrite it there .

Comment (behnazh-w) [reply src/macaron/malware_analyzer/pypi_heuristics/metadata/similar_projects.py:75]: If this code can obtain the maintainer info, rather than adding a new function, please update the `PyPIRegistry.get_maintainers_of_package` function so other heuristics can benefit from it.

Comment (AmineRaouane) [reply src/macaron/malware_analyzer/pypi_heuristics/metadata/similar_projects.py:75]: Should I change it here, or create a new PR for it?

Comment (behnazh-w) [reply src/macaron/malware_analyzer/pypi_heuristics/metadata/similar_projects.py:75]: Either works, although a separate PR is preferable.

**Review by behnazh-w — APPROVED** [review]
